client/lib/user/user.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * External Dependencies
+ */
+import { EventEmitter } from 'events';
+import { Language } from '../../languages';
+import { GravatarOptions } from './gravatar';
+import { URL } from '../../types';
+
+/**
+ * Takes an optional callback function. If no callback is passed in, it will return a promise.
+ **/
+export interface RequestHandler {
+	( fn: ( err: unknown, res: unknown, headers: unknown ) => unknown ): unknown;
+	(): Promise< unknown >;
+}
```

Comment:
we can probably be a little more specific here since people will be calling this and wanting to know what arguments the callback gets. it probably doesn't matter so don't let this block the PR…

```ts
type WPCOMError = {message: string};
type WPCOMHeaders = {[name: string]: string} & {status: number};

type WPCOMCallback = (error: WPCOMError | null, result: JSONSerializable | null, headers: WPCOMHeaders) => void;

interface User {
	sendVerificationEmail<F extends WPCOMCallback>(f: F): XMLHttpRequest;
	sendVerificationEmail(): Promise<void>;
}
```

as written we can pass a callback and expect a promise in the return at the same time and that type-checks because _fewer_ arguments is valid for when more arguments are possible. my modification will prevent passing
in a callback _and_ expecting a promise, though I couldn't figure out how to indicate that _only_ `error` _or_ `result` could have a value.

<img width="686" alt="Screen Shot 2020-02-26 at 12 12 36 AM" src="https://user-images.githubusercontent.com/5431237/75320855-d1ce8780-582c-11ea-82de-3ddb71830bd2.png">


[in the playground](http://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAUgygeQHJwgJwJYEMA2GBeWARjtALxRED2VpWAdlAD5QDOwm9A5s1PQK4BbIugDcAKFCQoAdQAKAYQQBZAKJo0VNFAoBvQRFassXCAC42HDNwC+EqdHlLlACQhYAJulY6ougNr0WAYW7JxcALqhVrZQAGR+7FjA-KwWAsLoduKS4I6KKgq4OERYAMYA1r4AFOiaaBZOKur1vAI4OAA0UGiG-DjAFvDIqJi4BMSkbf1dUAAW7l5oabIFroveAJQ6AHxQAG5UGB4S4tbA6ABm5dAAqqzofuIAkA-0HgBq6BiXGGXJGCo9FUgiwGBwAB4AGJQCAADwu7x8TWURQ6pUqO2qlwsUM2FgAGsoADIuYDAMAAJQgAEd+IZgBJXhB3l9ML9-sBAcDQeDqvioHJNIIMA8IYdjjsJDYcl4yjgsL0oGUgewoKl0BZ7mIchq0AA6N6fb4cgFAkFgnD8-XABb0ar83bK1W0CD6nBULjVADkYGFoog3s2m1OesNLON7L+Zp5lod2zIe10LxV9FYrvdnp9ACtUsAoFhlcUMRUg9KQ7qHgajWyftGuebeVbHYmns9U+nSJmvd7c2rC-90eVSxWbJsbXb406OxmPT36FQoH6qCKHj4AO4YW1FoeVVhBiviMM1k317kWvm1DRabq9Vj9YDdBaeLZO5PPAD0H6ggi00Gshw4PsEAeDuJTDjkzw-FAV71Ns77ti6XZzrBWgVs8MovNB1R3g+8EvIhaazlmuEDOhmGEZ2booc+SysKO462iyU6tjOyFZt6-z0Au+a9CkaCMMuq7QAsSqbtug7gZUB5AA)

Reply:
Nice, that's very satisfying to see in action 😂

Reply:
yeah you get autocomplete at least on `headers.status` too - I pulled that from the `wpcom-proxy-request` source. eventually it'd be best to just pull from there but that's an unreasonable ask for this PR